crates/module-system/sov-modules-api/src/state/accessors/concurrent_state_checkpoint.rs

```diff
@@ -0,0 +1,67 @@
+#![allow(dead_code)] // TODO: Remove this once the implementation is complete.
+use std::sync::Arc;
+
+use sov_rollup_interface::common::{RollupHeight, VisibleSlotNumber};
+use sov_state::{Namespace, SlotKey, SlotValue, StateGetter};
+
+use crate::{Spec, StateCheckpoint, TxChangeSet};
+
+/// An analogue of `StateCheckpoint` that can be safely written while concurrent reads are happening.
+pub struct ConcurrentStateCheckpoint<S: Spec> {
+    pub(super) storage: S::Storage,
+    pub(crate) uncomitted_changes: Option<Box<dyn StateGetter>>,
+    pub(crate) writes: Arc<concread::hashmap::HashMap<(SlotKey, Namespace), Option<SlotValue>>>,
+    pub(super) visible_slot_num: VisibleSlotNumber,
+    pub(super) rollup_height: RollupHeight,
+}
+
+impl<S: Spec> ConcurrentStateCheckpoint<S> {
+    /// Create a `ConcurrentStateCheckpoint` containing the same changes as the given `StateCheckpoint`.
+    pub fn from_state_checkpoint(mut state_checkpoint: StateCheckpoint<S>) -> Self {
+        let map = concread::hashmap::HashMap::new();
+        state_checkpoint.delta.commit_revertable_storage_cache();
+        let mut writer = map.write();
+        for (key, value) in state_checkpoint.delta.user_cache.get_writes() {
+            writer.insert((key.clone(), Namespace::User), value.cloned());
```

Comment:
Nit: Because key/value structures are cheap to clone:  Since we consume `state_checkpoint`, we could expose an API that also consumes `user_cache, kernel_cache, and accessory_writes`, avoiding the need to clone here.

Reply:
Done.